odl/util/phantom.py

```diff
@@ -398,6 +398,98 @@ def shepp_logan(space, modified=False):
     return phantom(space, ellipses)
 
 
+def submarine_phantom(discr, smooth=False):
+    """Return a 'submarine' phantom consisting in an ellipsoid and a box.
+
+    This phantom is used in [1]_ for shape-based reconstruction.
+
+    Parameters
+    ----------
+    discr : `Discretization`
+        Discretized space in which the phantom is supposed to be created
+    smooth : `bool`, optional
+        If `True`, the boundaries are smoothed out. Otherwise, the
+        function steps from 0 to 1 at the boundaries.
+
+    Returns
+    -------
+    phantom : `LinearSpaceVector`
+
+    References
+    ----------
+    .. [1] Oktem, Ozan. Mathematics of electron tomography. In:
+       Handbook of Mathematical Methods in Imaging. Scherzer, Otmar,
+       Ed. Springer, 2015, pp 937--1031.
+    """
+    if discr.ndim == 2:
+        if smooth:
+            return _submarine_phantom_2d_smooth(discr)
+        else:
+            return _submarine_phantom_2d_nonsmooth(discr)
+    else:
+        raise ValueError('Phantom only defined in 2 dimensions, got {}.'
+                         ''.format(discr.dim))
+
+
+def _submarine_phantom_2d_smooth(discr):
+    """Return a 2d smooth 'submarine' phantom."""
+
+    # Smaller value = more blurring
+    c = 20
+
+    def logistic(x):
+        return 1. / (1 + np.exp(-x))
+
+    def blurred_ellipse(x):
+        sq_ndist = np.zeros_like(x[0])
+        radii = (2, 0.7)
+        center = (0.5, -1)
+
+        for xi, rad, cen in zip(x, radii, center):
+            sq_ndist = sq_ndist + ((xi - cen) / rad) ** 2
+
+        out = np.sqrt(sq_ndist)
+        out -= 1
+        out *= (-c)
+        return logistic(out)
+
+    def blurred_rect(x):
+        out = np.ones_like(x[0])
+        xlower, xupper = (.3, -.5), (1.3, .5)
+        for xi, low, upp in zip(x, xlower, xupper):
+            out = out * logistic(c * (xi - low)) * logistic(c * (upp - xi))
+        return out
+
+    out = discr.element(blurred_ellipse)
+    out += discr.element(blurred_rect)
+    return out.ufunc.minimum(1)
+
+
+def _submarine_phantom_2d_nonsmooth(discr):
+    """Return a 2d nonsmooth 'submarine' phantom."""
+
+    def ellipse(x):
+        sq_ndist = np.zeros_like(x[0])
+        radii = (2, 0.7)
+        center = (0.5, -1)
+
+        for xi, rad, cen in zip(x, radii, center):
+            sq_ndist = sq_ndist + ((xi - cen) / rad) ** 2
+
+        return np.where(sq_ndist <= 1, 1, 0)
+
+    def rect(x):
+        out = np.ones_like(x[0])
+        xlower, xupper = (.3, -.5), (1.3, .5)
+        for xi, low, upp in zip(x, xlower, xupper):
+            out = out * np.where(np.logical_and(xi >= low, xi <= upp), 1, 0)
```

Comment:
Alternatively I would have to allocate a big array from the beginning and work with that. Here I start with the size of the first element, which is essentially one-dimensional, so operations are basically for free.